### PR TITLE
enhancement(config): Remove the watcher's coupling to SIGHUP / prepare for automatic Windows config reload

### PR DESCRIPTION
**Pull request by darklajid**
Right now the watcher is broken/unimplemented for Windows¹ and a large part of the
reason for that _might_ be how it works:

* the watcher subscribes to filesystem events (crate: notify²)
* the watcher raises a unix signal (SIGHUP) on detecting changes
* the signal handler (again, unix only) reacts to SIGHUP with an internal
  reload signal
* (vector magically reloads/compares the configs)

With these changes that unix signal dependency is removed

* the watcher clones a SignalTx to emit its very own signals
* the watcher sends an internal reload signal on detecting changes
* (vector magically reloads/compares the configs)

All of this _should_ also work on Windows, has nothing to do with unix signals anymore.

I slightly modified the tests to not just wait for any Ok, but to state explicitly what
signal we expect to receive.

As I am currently (and ironically..) on Linux, I haven't been able to verify the functionality on Windows yet, I'm trying to build a VM for that at the moment. I also expect fun challenges to build the dependencies on Windows.

For Linux, the tests remain green and this seems to be a transparent change to me: I don't think that affects any users, a SIGHUP works the same way it did before and just the filesystem change events don't go through that indirection anymore.

①: It fails outright/intentionally on Windows with an error pointing to https://github.com/vectordotdev/vector/issues/938, but it's unclear why this can't work
②: Notify officially supports native filesystem watches (and - as on all platforms - a polling fallback) on Windows

### CODE REVIEW

**Comment by darklajid** [comment]
Building on Windows has been a bit of a pain (arguably worse because I do it in a VM?). So far I have confirmed that this PR

* builds successfully
* the build above does what it should (i.e. `vector -w -c vector.toml` will pick up changes if the file is modified)

what is outstanding due to the insane amount of time it takes for me to compile stuff in this setup is a `cargo test watcher` run - that's currently building while I am sword-fighting on my office chair in the hallways.. Stiill, based on the observation of the `cargo build --release` binary from above I am already quite confident that this not only removes the whole weird SIGHUP dependency, but also magically allows Windows users to get hot-reloads.

**Comment by jszwedko** [comment]
Thanks for opening this @darklajid ! We'll review it shortly.

**Review by lukesteensen — APPROVED** [review]
This look sensible to me. Thanks @darklajid! And apologies for the delayed review.

File: changelog.d/support-config-watch-for-windows.enhancement.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Windows now supports the -w command line parameter just like every other
+platform and will reload the configuration files on any change
```
Comment (jszwedko) [inline changelog.d/support-config-watch-for-windows.enhancement.md:2]:
```suggestion
platform and will reload the configuration files on any change

authors: darklajid
```

This'll be dinged by CI without this line.

**Pull request merged**